partition/bootloader.go

```diff
@@ -109,208 +78,33 @@ func bootloaderImpl(p *Partition) (bootLoader, error) {
 	return nil, ErrBootloader
 }
 
-type bootloaderType struct {
-	partition *Partition
-
-	// each rootfs partition has a corresponding u-boot directory named
-	// from the last character of the partition name ('a' or 'b').
-	currentRootfs string
-	otherRootfs   string
-
-	// full path to rootfs-specific assets on boot partition
-	currentBootPath string
-	otherBootPath   string
-
-	// FIXME: this should /boot if possible
-	// the dir that the bootloader lives in (e.g. /boot/uboot)
-	bootloaderDir string
-}
-
-func newBootLoader(partition *Partition, bootloaderDir string) *bootloaderType {
-	// no root partition means we are in an all-snap system
-	if partition.rootPartition() == nil || partition.otherRootPartition() == nil {
-		// FIXME: once we no longer support snappy-ab remove
-		//        everything except bootloaderDir from the
-		//        bootloaderType struct
-		return &bootloaderType{
-			partition:     partition,
-			bootloaderDir: bootloaderDir,
-		}
-	}
-
-	// full label of the system {system-a,system-b}
-	currentLabel := partition.rootPartition().name
-	otherLabel := partition.otherRootPartition().name
-
-	// single letter description of the rootfs {a,b}
-	currentRootfs := string(currentLabel[len(currentLabel)-1])
-	otherRootfs := string(otherLabel[len(otherLabel)-1])
-
-	return &bootloaderType{
-		partition: partition,
-
-		currentRootfs: currentRootfs,
-		otherRootfs:   otherRootfs,
-
-		// the paths that the kernel/initramfs are loaded, e.g.
-		// /boot/uboot/a
-		currentBootPath: filepath.Join(bootloaderDir, currentRootfs),
-		otherBootPath:   filepath.Join(bootloaderDir, otherRootfs),
-
-		// the base bootloader dir, e.g. /boot/uboot or /boot/grub
-		bootloaderDir: bootloaderDir,
-	}
-}
-
-// FIXME:
-// - populate kernel if missing
-func (b *bootloaderType) SyncBootFiles(bootAssets map[string]string) (err error) {
-	for src, dst := range bootAssets {
-		if err := helpers.CopyIfDifferent(src, filepath.Join(b.bootloaderDir, dst)); err != nil {
+func markBootSuccessful(bootloader bootLoader) error {
+	// FIXME: we should have something better here, i.e. one write
+	//        to the bootloader environment only (instead of three)
+	//        We need to figure out if that is possible with grub/uboot
+	// (if we could also do atomic writes to the boot env, that would
```

Comment:
@niemeyer Yes, this probably sounds overly pessimistic. The comment tries to express that grub is not edited via a normal file but instead via a command line (grub-editenv). And uboot is edited via a special uboot.env file which is a normal file on a vfat partition. But we try to avoid anything that might move it on disk to avoid uboot falling over (this is why we rewrite it in place because we were bitten by bugs in the fat handling code in uboot). I am happy to rewrite the FIXME to make this cleaner.